components/blitz/src/omero/gateway/facility/DataManagerFacility.java

```diff
@@ -497,4 +572,242 @@ public void addImagesToDataset(SecurityContext ctx,
+     * @param file
+     *            The {@link File} to upload/attach
+     * @param mimetype
+     *            The mime type of the file (can be <code>null</code>)
+     * @param description
+     *            The description (can be <code>null</code>)
+     * @param namespace
+     *            The namespace (can be <code>null</code>)
+     * @param target
+     *            The {@link DataObject} to attach the file to (can be
+     *            <code>null</code>)
+     * @param callback
+     *            If no {@link Callback} is provided, the method will block
+     *            until the task is finished, otherwise the upload runs
+     *            asynchronously and the {@link Callback} handle gets notified
+     *            about the outcome
+     * @return The {@link FileAnnotationData} if no {@link Callback} handle was
+     *         provided, <code>null</code> otherwise
+     */
+    public FileAnnotationData attachFile(final SecurityContext ctx,
+            final File file, String mimetype, final String description,
+            final String namespace, final DataObject target, Callback callback) {
+        final String name = file.getName();
+        String absolutePath = file.getAbsolutePath();
+        final String path = absolutePath.substring(0, absolutePath.length()
+                - name.length());
+
+        final String mime;
+        if (mimetype == null)
+            mime = "application/octet-stream";
+        else
+            mime = mimetype;
+
+        final Callback cb = callback != null ? callback : new Callback();
+
+        Runnable r = new Runnable() {
+            @Override
+            public void run() {
+                RawFileStorePrx rawFileStore = null;
+                try {
+                    OriginalFile originalFile = new OriginalFileI();
+                    originalFile.setName(omero.rtypes.rstring(name));
+                    originalFile.setPath(omero.rtypes.rstring(path));
+                    originalFile.setSize(omero.rtypes.rlong(file.length()));
+                    final ChecksumAlgorithm checksumAlgorithm = new ChecksumAlgorithmI();
+                    checksumAlgorithm.setValue(omero.rtypes
+                            .rstring(ChecksumAlgorithmSHA1160.value));
+                    originalFile.setHasher(checksumAlgorithm);
+                    originalFile.setMimetype(omero.rtypes.rstring(mime));
+                    originalFile = (OriginalFile) saveAndReturnObject(ctx,
+                            originalFile);
+
+                    rawFileStore = gateway.getRawFileService(ctx);
+                    rawFileStore.setFileId(originalFile.getId().getValue());
+                    FileInputStream stream = new FileInputStream(file);
+                    long pos = 0;
+                    int rlen;
+                    byte[] buf = new byte[INC];
+                    ByteBuffer bbuf;
+                    while (((rlen = stream.read(buf)) > 0) && !cb.isCancelled()) {
```

Comment:
Much of this code exists in `omero.client`, though it doesn't include the callback. If the addition of a callback to `omero.client.upload` would make this reusable, that might well be worth it.